applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -0,0 +1,856 @@
+    {
+        if (distances[i] < 0.0)
+            rResult[NumNodes + i] =
+                GetGeometry()[i].GetDof(VELOCITY_POTENTIAL).EquationId();
+        else
+            rResult[NumNodes + i] =
+                GetGeometry()[i].GetDof(AUXILIARY_VELOCITY_POTENTIAL).EquationId();
+    }
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::GetDofListNormalElement(DofsVectorType& rElementalDofList) const
+{
+    for (unsigned int i = 0; i < NumNodes; i++)
+        rElementalDofList[i] = GetGeometry()[i].pGetDof(VELOCITY_POTENTIAL);
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::GetDofListKuttaElement(DofsVectorType& rElementalDofList) const
+{
+    const auto& r_geometry = this->GetGeometry();
+    // Kutta elements have only negative part
+    for (unsigned int i = 0; i < NumNodes; i++)
+    {
+        if (!r_geometry[i].GetValue(TRAILING_EDGE))
+            rElementalDofList[i] = r_geometry[i].pGetDof(VELOCITY_POTENTIAL);
+        else
+            rElementalDofList[i] = r_geometry[i].pGetDof(AUXILIARY_VELOCITY_POTENTIAL);
+    }
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::GetDofListWakeElement(DofsVectorType& rElementalDofList) const
+{
+    array_1d<double, NumNodes> distances;
+    GetWakeDistances(distances);
+
+    // Positive part
+    for (unsigned int i = 0; i < NumNodes; i++)
+    {
+        if (distances[i] > 0)
+            rElementalDofList[i] = GetGeometry()[i].pGetDof(VELOCITY_POTENTIAL);
+        else
+            rElementalDofList[i] = GetGeometry()[i].pGetDof(AUXILIARY_VELOCITY_POTENTIAL);
+    }
+
+    // Negative part - sign is opposite to the previous case
+    for (unsigned int i = 0; i < NumNodes; i++)
+    {
+        if (distances[i] < 0)
+            rElementalDofList[NumNodes + i] = GetGeometry()[i].pGetDof(VELOCITY_POTENTIAL);
+        else
+            rElementalDofList[NumNodes + i] =
+                GetGeometry()[i].pGetDof(AUXILIARY_VELOCITY_POTENTIAL);
+    }
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::CalculateLeftHandSideNormalElement(
+    MatrixType& rLeftHandSideMatrix, const ProcessInfo& rCurrentProcessInfo)
+{
+    if (rLeftHandSideMatrix.size1() != NumNodes || rLeftHandSideMatrix.size2() != NumNodes)
+        rLeftHandSideMatrix.resize(NumNodes, NumNodes, false);
+    rLeftHandSideMatrix.clear();
+
+    ElementalData<NumNodes, Dim> data;
+
+    // Calculate shape functions
+    GeometryUtils::CalculateGeometryData(GetGeometry(), data.DN_DX, data.N, data.vol);
+
+    const double density = ComputeDensity(rCurrentProcessInfo);
+    const double DrhoDu2 = ComputeDensityDerivative(density, rCurrentProcessInfo);
+
+    // Computing local velocity
+    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
```

Comment:
```suggestion
    const auto& r_free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
```
I don't think you need a copy in this case.

Reply:
I agree that here reference is preferred, however I would not recommend using auto for velocity vectors inside this element since it obscures the size of the vector and can complicate the development later on. In 2D the velocity vector has sometimes two components and sometimes three.